### ISSUE TITLE
Firing triggers from OnEntry/OnExit actions.

### ISSUE THREAD

**Issue by ianwall** [dotnet-state-machine/stateless#11]
This is something I do quite frequently - an OnEntry action either directly or indirectly results a call to Fire() that takes the state machine to a new state.

The problem is that because the call to Fire immediately causes a new transition, any code after the call to Fire in the OnEntry action is called _after_ the state machine enters the new state. Here's an example:

```
class Program
{
    enum Trigger
    {
        T1, T2, T3
    }

    enum State
    {
        S1, S2, S3
    }

    static void Main(string[] args)
    {
        var check = true;

        var sm = new StateMachine<State, Trigger>(State.S1);

        sm.Configure(State.S1)
            .Permit(Trigger.T1, State.S2)
            .OnEntry(() => Console.WriteLine("State S1 OnEntry action"));

        sm.Configure(State.S2)
            .Permit(Trigger.T2, State.S3)
            .Permit(Trigger.T3, State.S1)
            .OnEntry(() =>
                {
                    Console.WriteLine("State S2 OnEntry action started");
                    if (check)
                        sm.Fire(Trigger.T2);
                    else
                        sm.Fire(Trigger.T3);
                    Console.WriteLine("State S2 OnEntry action complete");
                });

        sm.Configure(State.S3)
            .Permit(Trigger.T1, State.S1)
            .OnEntry(() => Console.WriteLine("State S3 OnEntry action"));

        sm.Fire(Trigger.T1);
    }
}
```

This results in the output:

State S2 OnEntry action started
State S3 OnEntry action
State S2 OnEntry action complete

I.e. S2 completes after S3.

I have a simple fix for this that add an internal queue to the state machine so that processing of triggers is carried in strict order.

Do you want the fix?

**Comment by StevenTCramer** [comment]
Ian I would love to see this fixed.

**Comment by HenningNT** [comment]
I like to wrap the state machine in an Active Object, that provides an Interface to the world. The outside would not deal with the state machine, but use the public methods to post triggers in a queue. A timer empties the queue at regular intervals, so that all triggers are dealt with in order and atomically.

All trigger handlers also uses the queue, they don't Fire triggers. So I would have something like this for the OnEntry:

            .OnEntry(() =>
                {
                    Console.WriteLine("State S2 OnEntry action started");
                    if (check)
                        _eventQueue.Enqueue(Trigger.T2);
                    else
                        _eventQueue.Enqueue(Trigger.T3);
                    Console.WriteLine("State S2 OnEntry action complete");
                });

**Comment by ianwall** [comment]
I understand wrapping the state machine in an Active Object, and using a timer, but I want my state machines to transition immediately one state action has finished, if there is an event in the pipeline. I do not want to wait for a timer to fire. Also it saves having a timer resource created for each instance of the machine - and I sometimes have many instances.

**Comment by SirRufo** [comment]
**One** trigger should only cause **one** transition or none. It is just a trigger and not a storyteller. If you want to have a dynamic behavior on the trigger use PermitDynamic

**Comment by HenningNT** [comment]
Ian, here's how I do the Queue handling:

        private void TickTimerCallback(object state)
        {
            Trigger trigger;

            while (_lifoQueue.TryPop(out trigger))
            {
                ProcessTrigger(trigger);
            }

            while (MessageQueue.TryDequeue( out trigger))
            {
                ProcessTrigger(trigger);

                while (_lifoQueue.TryPop(out trigger))
                {
                    ProcessTrigger(trigger);
                }
            }

            _tickTimer.Change(QueueInterval, Timeout.Infinite);
        }

It will process Triggers as long as there is a Trigger in the pipeline.
I found that some Triggers must be processed before others, so I have two Queues; ConcurrentQueue for FIFO, and ConcurrentStack for LIFO.

**Comment by jklemmack** [comment]
Ian - did you ever merge the code?  I would really like this feature too, since it helps break apart sub-processing elements.  If not, I can fork and would love a PR or dump of your changes.

**Comment by ianwall** [comment]
Hi Johann, no I didn't merge the code. I'll try and do a PR (a bit of a novelty for me).

**Comment by ianwall** [comment]
Ok, I've created a  PR.

**Comment by HenningNT** [comment]
I looked  at the PR, but I haven't tried it yet. From what I can tell it will proccess events FIFO, in the same order as they are added to the queue. But sometimes events have to be proccessed LIFO instead of FIFO.
Also, maybe use BlockingCollection or ConcurrentQueue to avoid concurrency issues?

**Comment by jklemmack** [comment]
Sounds to me like a thread-safe list or array would fit the bill.  You can
dequeue off the end, and either enqueue to do FIFO, or Push for LIFO.  It
looks like the access points are all internal, so easy to guard them in a
thread-safe manner.

Pretty nonstandard behavior however, so huzzah for docs.  Perhaps guard the
behavior with a switch at the StateMachine level so it isn't a breaking
change?

**Comment by nblumhardt** [comment]
@HenningNT @jklemmack just a note, Stateless is not concurrent/thread-safe at all, so I don't think the collection type here needs to be a concurrent one. Cheers!

**Comment by nblumhardt** [comment]
I have thought about this one some more, and I'm reaching the conclusion that either we make this change and accept the PR, or we forego re-entrancy altogether.

With the status quo, because a state and its superstate may have multiple entry/exit handlers defined, calling back into Fire() from a handler means the later handlers may fire with the statemachine in a completely unrelated state. From this angle, making the change to FIFO-run-to-completion seems like a step forward, even if there's no perfect option to cover all cases.

**Comment by nblumhardt** [comment]
Merged in #47.
